openpype/plugins/publish/extract_slate_global.py

```diff
@@ -0,0 +1,538 @@
+
+from openpype.lib import (
+    get_oiio_tools_path,
+    get_ffprobe_streams,
+    get_chrome_tool_path,
+    run_subprocess,
+)
+from openpype.pipeline import publish
+
+
+def timecode_to_frames(timecode, framerate):
+    rt = otio.opentime.from_timecode(timecode, framerate)
+    return int(otio.opentime.to_frames(rt))
+
+
+def frames_to_timecode(frames, framerate):
+    rt = otio.opentime.from_frames(frames, framerate)
+    return otio.opentime.to_timecode(rt)
+
+
+def frames_to_seconds(frames, framerate):
+    rt = otio.opentime.from_frames(frames, framerate)
+    return otio.opentime.to_seconds(rt)
+
+
+class SlateCreator:
+    """Class that formats and renders an html template to use as a slate.
+
+    Html renders as sRGB (or Rec.709 if you want a little gamma boost), so
+    rendering will happen in .png as intermediate file, while then converted
+    to dest cspace using oiio --colorconvert attributes syntax.
+
+    """
+
+    def __init__(
+        self,
+        template_path,
+        resources_path="",
+        staging_dir="",
+        staging_subfolder="",
+        data=None,
+    ):
+        self.data = data.copy() if data else {}
+        self.log = logging.getLogger(self.__class__.__name__)
+
+
+        self.set_template_paths(template_path, resources_path=resources_path)
+        self.set_staging_dir(staging_dir, subfolder=staging_subfolder)
+        self.read_template()
+        self.task_filter = []
+
+    def set_template_paths(self, template_path, resources_path=""):
+        """Set path of HTML slate template file and CSS resources."""
+
+        self.template_path = os.path.normpath(template_path)
+        self.log.debug(
+            "Using Slate template path: '%s'", self.template_path
+        )
+
+        if resources_path:
+            self.resources_path = os.path.normpath(resources_path)
+            self.log.debug(
+                "Using Slate resources path: '%s'", self.resources_path
+            )
+        else:
+            basedir = os.path.basename(self.template_path)
+            self.resources_path = os.path.normpath(basedir)
+            self.log.debug(
+                "No Slate resources path specified, using: '%s'", self.resources_path
+            )
+
+    def set_resolution(self, width, height):
+        self.data["resolution_width"] = width
+        self.data["resolution_height"] = height
+        self.log.debug("Slate resolution set to: '%sx%s'", width, height)
+
+    def set_staging_dir(self, path="", subfolder=""):
+        if path:
+            staging_dir = os.path.normpath(path)
+        else:
+            staging_dir = tempfile.gettempdir()
+
+        if subfolder:
+            staging_dir = os.path.join(staging_dir, subfolder)
+
+        if not os.path.exists(staging_dir):
+            os.makedirs(staging_dir)
+
+        self.staging_dir = staging_dir
+        self.log.debug("Staging dir set to: '%s'", staging_dir)
+
+    def read_template(self, template_path=None, resources_path=None):
```

Comment:
Is it weird maybe that the class `__init__` allows to take in a template path and resources path to initialize but then these methods also do the same? I feel like with an instantiated object it makes more settings that it either takes the settings from the instance **or** that this method re-initializes the object to that new template, however this function does not seem to override the `self.template_path` or `self.resources_path` if it were supplied.

Should `read_template` either take the values from `self` OR should it be the one that is also storing/assigning it to `self` instead?

Reply:
Yeah I agree and I had the same thought while refactoring Max's code in a few other places with some of the variables stored on the instance. I did remove a few of them but in some places like this one it felt like someone could benefit from being able to override the template used after the SlateCreator object had been created... although I'm usually more of the philosophy of starting as simple as possible and only expand logic if needed.

So long story short, I'd be happy on removing the args from this function and just using the ones initialized on the instance, what do you think?